### PR TITLE
Feature - add HashiCorp Vault secret provider

### LINKED ISSUE #159: Provide HashiCorp Vault secret provider
**Is your feature request related to a problem? Please describe.**
Provide HashiCorp Vault secret provider

### PR DESCRIPTION
**Pull request by stijnmoreels**
Added a new project (NuGet package) with a secret provider that will interact wtih the KeyValue secret of a HashiCorp Vault.
This feature contains:
- The `ISecretProvider` implementation itself 
- User-friendly extensions to register in the secret store for:
    - UserPass authentication
    - Kubernetes authentication
    - Custom configuration
- Unit tests for all components
- Integration test that interact with a fresh HashiCorp Vault instance running in 'dev server' mode for:
    - UserPass authentication
- Feature docs update with examples of how this secret provider can be added to the secret store, with examples of what is required and by default provided and can be overridden.

Relates to #159

### CODE REVIEW

**Comment by tomkerkhove** [comment]
> Not yet created integration test for Kuberetes authentication bc that would require me to create a Kubernetes service account (extra costs per month) and maybe not the worth the cost for that?
> Of course if there's a way that this can be up shortcutted, and there's a way to test this external call by setting up a test server that validates the JWT token for us... we could maybe go that route.

Let's add a dedicated issue where we deploy a container to a [Kind cluster](https://kind.sigs.k8s.io/) and run against that.

File: src/Arcus.Security.Providers.HashiCorp/VaultKeyValueSecretEngineVersion.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Arcus.Security.Providers.HashiCorp
+{
+    /// <summary>
+    /// Represents the API version of the client engines when interacting with the HashiCorp Vault in the <see cref="HashiCorpSecretProvider"/>.
```
Comment (tomkerkhove) [inline src/Arcus.Security.Providers.HashiCorp/VaultKeyValueSecretEngineVersion.cs:4]: Maybe link to the official docs for this?

**Review by tomkerkhove — APPROVED** [review]
LGTM, the only suggestion I would have is to move out the Kubernetes auth unless you are certain that it already works fine?

**Comment by stijnmoreels** [comment]
> LGTM, the only suggestion I would have is to move out the Kubernetes auth unless you are certain that it already works fine?

I can honestly say I don't know anything for certain 😅.

File: build/templates/download-hashicorp-vault.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        echo "##vso[task.complete result=Failed;]"
+      fi
+    env:
+      TARGET_FOLDER: ${{ parameters.targetFolder }}
+      VERSION: ${{ parameters.version }}
+      VARIABLE_NAME: ${{ parameters.vaultBinVariableName }}
+    displayName: 'Check for required parameters in YAML template'
+  - powershell: |
+      # New-Item $env:TARGET_FOLDER
+      $vault_zip = "vault_$($env:VAULT_VERSION)_linux_amd64.zip"
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply build/templates/download-hashicorp-vault.yml:27]: Ah, yes, was something to do with availability in inline scripts but I think that's only for bash scripts. I think indeed that we can use the parameters directly. We'll check that.

File: src/Arcus.Security.Providers.HashiCorp/VaultKeyValueSecretEngineVersion.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.Security.Providers.HashiCorp/VaultKeyValueSecretEngineVersion.cs:4]: Yes, was thinking maybe we should provide more links to the HashiCorp Vault docs throughout the source code and docs.

**Comment by tomkerkhove** [comment]
Good enough!

File: build/templates/download-hashicorp-vault.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
+parameters:
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply build/templates/download-hashicorp-vault.yml:1]: Hmm, I don't find it in the docs that they validate for `null` strings when there's no default present?

**Pull request merged**